File: cpp/example_code/redshift/authorize_cluster_access.cpp
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
#include <aws/core/Aws.h>
#include <aws/ec2/EC2Client.h>
#include <aws/ec2/model/AuthorizeSecurityGroupIngressRequest.h>
#include <iostream>

/**
 * Enable access to Amazon Redshift clusters
 *
 * Defines a security group inbound rule for the default VPC. The rule
 * enables access to Redshift clusters by IP addresses referenced in the
 * ipAddress argument. To define the rule, EC2 permissions are required.
 */
bool AuthorizeClusterAccess(const Aws::String & ipAddress)
{
    Aws::EC2::EC2Client ec2;
    Aws::EC2::Model::AuthorizeSecurityGroupIngressRequest ec2_req;

    // Initialize request parameters
    ec2_req.SetGroupName("default");
    ec2_req.SetIpProtocol("tcp");
    ec2_req.SetFromPort(5439);		// Default Redshift port
    ec2_req.SetToPort(5439);		// Default Redshift port
    ec2_req.SetCidrIp(ipAddress);

    // Define the inbound rule
    auto outcome = ec2.AuthorizeSecurityGroupIngress(ec2_req);

    if (!outcome.IsSuccess())
    {
        std::cerr << "Error allowing cluster access. " << 
            outcome.GetError().GetMessage() << std::endl;
        return false;
    }
    return true;
}

/**
 * Exercise AuthorizeClusterAccess()
 */
int main(int argc, char **argv)
{
    Aws::SDKOptions options;
    Aws::InitAPI(options);
    {
        // Set these configuration values before running the program
        // The demo's ipAddress setting allows access from any computer. This 
        // is reasonable for demonstration purposes, but is not appropriate in
        // a production environment.
        Aws::String ipAddress = "0.0.0.0/0";

        if (!AuthorizeClusterAccess(ipAddress))
        {
            return 1;
        }

        std::cout << "Enabled access to Amazon Redshift clusters." << std::endl;
    }
    Aws::ShutdownAPI(options);
    return 0;
}

File: python/example_code/redshift/authorize_cluster_access.py
# Copyright 2010-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
# This file is licensed under the Apache License, Version 2.0 (the "License").
# You may not use this file except in compliance with the License. A copy of
# the License is located at
#
# http://aws.amazon.com/apache2.0/
#
# This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
# CONDITIONS OF ANY KIND, either express or implied. See the License for the
# specific language governing permissions and limitations under the License.

# snippet-sourcedescription:[authorize_cluster_access.py demonstrates how to enable access to Amazon Redshift clusters.]
# snippet-service:[redshift]
# snippet-keyword:[Python]
# snippet-sourcesyntax:[python]
# snippet-keyword:[Amazon Redshift]
# snippet-keyword:[Code Sample]
# snippet-sourcetype:[full-example]
# snippet-sourcedate:[2019-01-28]
# snippet-sourceauthor:[AWS]
# snippet-start:[redshift.python.authorize_cluster_access.complete]

import boto3
from botocore.exceptions import ClientError


def authorize_cluster_access(IpAddress='0.0.0.0/0'):
    """Enable access to Amazon Redshift clusters

    Defines a security group inbound rule for the default VPC. The rule
    enables access to Redshift clusters by IP addresses referenced in the
    IpAddress argument. To define the rule, EC2 permissions are required.

    :param IpAddress: string; IP addresses to authorize access to Redshift
    clusters. Default: '0.0.0.0/0' allows access from any computer, which is
    reasonable for demonstration purposes, but is not appropriate in a
    production environment.
    :return: True if cluster access is enabled, else False
    """

    ec2_client = boto3.client('ec2')

    # Redshift uses port 5439 by default. If Redshift was configured to use
    # a different port, specify the FromPort= and ToPort= arguments accordingly.
    try:
        ec2_client.authorize_security_group_ingress(GroupName='default',
                                                    IpProtocol='tcp',
                                                    FromPort=5439,
                                                    ToPort=5439,
                                                    CidrIp=IpAddress)
    except ClientError as e:
        print(f'ERROR: {e}')
        return False
    return True
# snippet-end:[redshift.python.authorize_cluster_access.complete]


def main():
    """Test authorize_cluster_access()"""
    if not authorize_cluster_access():
        print('FAIL: authorize_cluster_access()')
        exit(1)


if __name__ == '__main__':
    main()
